core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -1032,7 +1031,7 @@ private AcquiredRecords acquireNewBatchRecords(String memberId, long firstOffset
                 startOffset = firstOffset;
             }
             endOffset = lastOffset;
-            nextFetchOffset = nextOffset;
+            findNextFetchOffset.set(true);
```

Comment:
I don't think this is the right way to fix the issue as this change always enforces the complete iteration in `nextFetchOffset` which should not be required generally.